### PR TITLE
Transaction publication

### PR DESCRIPTION
**Pull request by da-kami**
- [x] Publish lock tx

No monitoring yet, decided to get this in and add the monitoring in a separate PR. (smaller chunks...)

Published contracts:

https://mempool.space/testnet/tx/a96a4f8795f9f7c06950f41986347caffe1e44bef4bbf3a67c905e4ac5870c57
and
https://mempool.space/testnet/tx/bbaf58bacdbc1946d851ecbc97fe020f0068d1985e6970218432e8f4d244d74a

### CODE REVIEW

File: daemon/src/maker_cfd_actor.rs [hunk 1]
```diff
@@ -266,9 +266,16 @@ pub fn new(
...
                             .unwrap();
 
-                        // TODO: Publish on chain and only then transition to open - this might
-                        // require saving some internal state to make sure we are able to monitor
-                        // the publication after a restart
+                        let sigend_tx_lock = wallet.sign(dlc.lock).unwrap();
+                        match wallet.try_broadcast_transaction(sigend_tx_lock).unwrap() {
+                            Ok(txid) => println!("Transaction sent to chain with txid {}", txid),
+                            Err(_already_in_chain) => {
+                                println!("Thanks for paying the transaction fee Mr. maker")
```
Comment (thomaseizinger) [inline daemon/src/maker_cfd_actor.rs]:
```suggestion
                                println!("Thanks for paying the transaction fee Mr. taker")
```

File: daemon/src/maker.rs [hunk 2]
```diff
@@ -123,7 +124,7 @@ async fn main() -> Result<()> {
 
                 let (cfd_maker_actor, cfd_maker_actor_inbox) = maker_cfd_actor::new(
                     db,
-                    wallet,
+                    Arc::new(wallet),
```
Comment (thomaseizinger) [inline daemon/src/maker.rs]: The `Wallet` has an `Arc` inside yeah? No need to `Arc` it again then :)

File: daemon/src/maker_cfd_actor.rs [hunk 3]
```diff
@@ -266,9 +270,19 @@ pub fn new(
...
-                        // require saving some internal state to make sure we are able to monitor
-                        // the publication after a restart
+                        match wallet
+                            .try_broadcast_transaction(dlc.lock.extract_tx())
+                            .await
+                            .unwrap()
+                        {
+                            Ok(txid) => println!("Transaction sent to chain with txid {}", txid),
+                            Err(_already_in_chain) => {
+                                println!("Thanks for paying the transaction fee Mr. taker")
```
Comment (thomaseizinger) [inline daemon/src/maker_cfd_actor.rs]: As we discovered, this is wrong :D

File: daemon/src/model/cfd.rs [hunk 4]
```diff
@@ -511,6 +511,8 @@ pub struct Dlc {
     pub revocation: SecretKey,
     pub publish: SecretKey,
 
+    /// The lock transaction that already contains the signature of the other party.
+    /// Each side now just has to sign it and can then publish the tx.
```
Comment (thomaseizinger) [inline daemon/src/model/cfd.rs]: Is that still true? I think it is fully signed now right?

File: daemon/src/setup_contract_actor.rs [hunk 5]
```diff
@@ -127,11 +130,22 @@ pub fn new(
...
+        // TODO: In case we sign+send but never receive (the signed lock_tx from the other party) we
+        // need some fallback handling (after x time) to spend the outputs in a different way so the
+        // other party cannot hold us hostage
+
         Dlc {
             identity: sk,
             revocation: rev_sk,
             publish: publish_sk,
-            lock: lock_tx,
+            lock: signed_lock_tx,
```
Comment (thomaseizinger) [inline daemon/src/setup_contract_actor.rs]: Can we just do `signed_lock_tx.extract_tx()` here and move from `Psbt` to `Transaction`?

File: daemon/src/wallet.rs [hunk 6]
```diff
@@ -61,4 +74,88 @@ impl Wallet {
...
+
+        wallet
+            .sign(
+                &mut psbt,
+                SignOptions {
+                    trust_witness_utxo: true,
+                    ..Default::default()
+                },
+            )
+            .context("could not sign lock tx")?;
```
Comment (thomaseizinger) [inline daemon/src/wallet.rs]: Nit: We no longer know that is really the lock tx in here.

File: daemon/src/wire.rs [hunk 7]
```diff
@@ -137,3 +160,8 @@ impl From<CfdTransactions> for Msg1 {
         }
     }
 }
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Msg2 {
+    pub signed_lock: PartiallySignedTransaction,
```
Comment (thomaseizinger) [inline daemon/src/wire.rs]: Ideally we would be using the psbt binary encoding here but that is not super important.

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Few nits, otherwise good.

File: daemon/src/maker_cfd_actor.rs (hunk 3 above)
Comment (da-kami) [reply daemon/src/maker_cfd_actor.rs]: 💩

File: daemon/src/maker_cfd_actor.rs [hunk 8]
```diff
@@ -175,7 +175,9 @@ pub fn new(
                         let cfd = load_cfd_by_order_id(order_id, &mut conn).await.unwrap();
                         let margin = cfd.margin().unwrap();
 
-                        let maker_params = wallet.build_party_params(margin, pk).unwrap();
+                        let maker_params = wallet.build_party_params(margin, pk).await.unwrap();
+
+                        let wallet = wallet.clone();
```
Comment (thomaseizinger) [inline daemon/src/maker_cfd_actor.rs]: That could in theory just be done in the function call to `new`.

**Review by thomaseizinger — APPROVED** [review]

File: daemon/src/wire.rs (hunk 7 above)
Comment (da-kami) [reply daemon/src/wire.rs]: I applied the same TODO as for `Msg0` 🤪

**Pull request merged**